File: recipes/subms-stats/rust/src/jitter.rs
//! Jitter score: was the measurement *environment* stable across the
//! run? Behind the `jitter` Cargo feature (on by default).

/// Jitter score: coefficient of variation of the per-window mean
/// across non-overlapping 32-sample windows. Clamped to `[0.0, 1.0]`.
/// Returns 0 when there are fewer than 64 samples (two windows).
///
/// The score answers: "did the *measurement* noise vary across the
/// run?" A high jitter score doesn't mean the algorithm is slow - it
/// means the numbers shifted under our feet (GC, NUMA migration, CPU
/// thermal throttling, OS scheduler preemption). Separate signal from
/// the per-stage p99 tail.
pub fn jitter_score(samples: &[u64]) -> f64 {
    const WIN: usize = 32;
    if samples.len() < WIN * 2 {
        return 0.0;
    }
    let windows = samples.len() / WIN;
    let mut means = Vec::with_capacity(windows);
    for w in 0..windows {
        let start = w * WIN;
        let slice = &samples[start..start + WIN];
        let sum: u64 = slice.iter().sum();
        means.push(sum as f64 / WIN as f64);
    }
    let grand_mean: f64 = means.iter().sum::<f64>() / means.len() as f64;
    if grand_mean <= 0.0 {
        return 0.0;
    }
    let variance: f64 =
        means.iter().map(|m| (m - grand_mean).powi(2)).sum::<f64>() / means.len() as f64;
    let cv = variance.sqrt() / grand_mean;
    cv.clamp(0.0, 1.0)
}

#[cfg(test)]
#[path = "jitter_tests.rs"]
mod tests;

File: recipes/subms-stats/java/src/main/java/com/submillisecond/recipes/stats/Jitter.java
package com.submillisecond.recipes.stats;

/**
 * Measurement-rig stability indicator: did the noise floor shift
 * across the run? Independent of the algorithm's own tail latency.
 *
 * <p>Byte-equivalent to {@code subms_stats::jitter} in the Rust
 * sibling.
 */
public final class Jitter {
    private Jitter() {}

    private static final int WIN = 32;

    /**
     * Jitter score: coefficient of variation of the per-window mean
     * across non-overlapping 32-sample windows. Clamped to
     * {@code [0.0, 1.0]}. Returns 0 when there are fewer than 64
     * samples (two windows).
     *
     * <p>A high jitter score doesn't mean the algorithm is slow - it
     * means the numbers shifted under your feet (GC, NUMA migration,
     * CPU thermal throttling, OS scheduler preemption).
     */
    public static double jitterScore(long[] samples) {
        if (samples.length < WIN * 2) return 0.0;
        int windows = samples.length / WIN;
        double[] means = new double[windows];
        for (int w = 0; w < windows; w++) {
            long sum = 0;
            int start = w * WIN;
            for (int i = 0; i < WIN; i++) sum += samples[start + i];
            means[w] = (double) sum / WIN;
        }
        double grand = 0;
        for (double m : means) grand += m;
        grand /= means.length;
        if (grand <= 0.0) return 0.0;
        double variance = 0;
        for (double m : means) {
            double d = m - grand;
            variance += d * d;
        }
        variance /= means.length;
        double cv = Math.sqrt(variance) / grand;
        return Math.max(0.0, Math.min(1.0, cv));
    }
}
